test/hotspot/jtreg/compiler/loopopts/InvariantCodeMotionReassociateCmp.java

```diff
@@ -0,0 +1,315 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.loopopts;
+
+import compiler.lib.ir_framework.*;
+import jdk.test.lib.Asserts;
+
+/*
+ * @test
+ * @bug 8323220
+ * @summary Test loop invariant code motion for cmp nodes through reassociation
+ * @library /test/lib /
+ * @run driver compiler.c2.loopopts.InvariantCodeMotionReassociateCmp
+ */
+public class InvariantCodeMotionReassociateCmp {
+    private static final int size = 500;
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @DontInline
+    private void blackhole() {}
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_I, "1"})
+    public int[] equalsAddInt(int inv1, int inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
+                break;
+            }
+        }
+        return new int[]{i, inv1, inv2};
+    }
+
+    @Check(test = "equalsAddInt")
+    public void checkEqualsAddInt(int[] returnValue) {
+        if (returnValue[0] != size && returnValue[0] + returnValue[1] != returnValue[2]) {
+            throw new RuntimeException("Illegal reassociation: i=" + returnValue[0] + ", inv1=" + returnValue[1]
+                    + ", inv2=" + returnValue[2]);
+        }
+    }
+
+    @Test
+    @Arguments({Argument.RANDOM_EACH, Argument.RANDOM_EACH})
+    @IR(counts = {IRNode.SUB_L, "1"})
+    public long[] equalsAddLong(long inv1, long inv2) {
+        int i = 0;
+        for (; i < size; ++i) {
+            // Reassociate to `inv2 - inv1 == i`
+            if (inv1 + i == inv2) {
+                blackhole();
+                break;
+            }
+        }
+        return new long[]{i, inv1, inv2};
+    }
+
+    @Check(test = "equalsAddLong")
+    public void checkEqualsAddLong(long[] returnValue) {
+        if (returnValue[0] != size && returnValue[0] + returnValue[1] != returnValue[2]) {
+            throw new RuntimeException("Illegal reassociation: i=" + returnValue[0] + ", inv1=" + returnValue[1]
+                    + ", inv2=" + returnValue[2]);
```

Comment:
Sadly, you cannot reuse `@Check` methods for multiple `@Test` methods. But you can have a utility method that you invoke from multiple `@Check` methods. That would reduce the code-duplication.